### ISSUE TITLE
http v.1.0 requests slips by access_by_lua_file

### ISSUE THREAD

**Issue by JarleB** [openresty/openresty#154]
How to reproduce:

nginx.conf for  http://nginx-front-end-proxy: 
```
location /  {
  default_type text/html;
  access_by_lua_file '/files/auth.lua';
  proxy_pass http:/backend-service;
}
```

/files/auth.lua:
```
local template = require "resty.template"
local message = "Enter password"
template.render("login.html", { message = message })
```

The following request will access http:/backend-service 

curl -0   http://nginx-front-end-proxy

Expected behavior: 

Rendered login.html is returned

**Comment by agentzh** [comment]
@JarleB I failed to reproduce the problem with the following minimal and standalone example:

```nginx
location / {
    access_by_lua_block {
        ngx.exit(403)
    }
}
```

Sending an HTTP 1.0 request:

```console
$ curl -0 localhost:8080/
<html>
<head><title>403 Forbidden</title></head>
<body bgcolor="white">
<center><h1>403 Forbidden</h1></center>
<hr><center>nginx/1.9.7 (no pool)</center>
</body>
</html>
```

It's exactly what we'd expect.

I guess you just forgot to call `ngx.exit()` in your `access_by_lua*` Lua handler to short-circuit the request processing.

**Comment by JarleB** [comment]
@agentzh  Thank you. Seems I had wrong assumptions about this. I'm still a bit puzzled why v1.0 and v1.1 requests is processed differently though?

**Comment by agentzh** [comment]
HTTP 1.0 is special since it lacks many features. If you find anything unusual, please provide a minimal and standalone example that everyone can easily run and reproduce the oddity you are talking about. Thank you.

**Comment by marcusramberg** [comment]
@agentzh seems that the problem is that with http 1.1 when the body is sent (template.render()), proxy_pass will not resend the body, but with http 1.0 it will overwrite the previously rendered template body. This seems like a gotcha, and I'm not sure why it should be different for http 1.0.

**Comment by agentzh** [comment]
@marcusramberg Maybe you should turn off the lua_http10_buffering directive in your nginx.conf?

https://github.com/openresty/lua-nginx-module#lua_http10_buffering

**Comment by agentzh** [comment]
@marcusramberg Because I don't maintain the lua-resty-template library, I don't quite understand your description above. Better derive a minimal and standalone example so that I can follow much more easily :)

**Comment by marcusramberg** [comment]
Yeah, that would probably fix my issue. Seems like the default for that setting should have been off though. just my 5c. Leads to very unexpected results.

Marcus

**Comment by agentzh** [comment]
@marcusramberg If you want to terminate the processing in your access by lua handler, then you should really call `ngx.exit(200)` (or something like that) in your Lua code. See

https://github.com/openresty/lua-nginx-module#ngxexit

**Comment by marcusramberg** [comment]
I completely agree that's the best solution, but most people won't be doing http 1.0 testing, so having it buffer differently from http 1.1 by default seems like the wrong choice which can lead to security holes. lua template uses ngx.print, and I have an example on my work computer I'll add tomorrow which illustrates the difference. 

I'm a bit surprised, as the http10_buffering documentation seems to indicate buffering will be disabled if you use ngx.print and write a body.

Marcus

**Comment by agentzh** [comment]
@marcusramberg HTTP 1.0 buffering is on by default because HTTP 1.0 keep-alive requires that feature.

**Comment by agentzh** [comment]
@marcusramberg You are *required* to call `ngx.exit()` to terminate the request handling if you generate response in the access phase (or rewrite phase). No matter whether HTTP 1.0 buffering is on or off.

**Comment by agentzh** [comment]
@marcusramberg `ngx.print` does not affect HTTP 1.0 buffering at all. Only the appearance of the `Content-Length` response header does.

**Comment by marcusramberg** [comment]
@agentzh From your documentation link above: 

```
If the Lua code explicitly sets a Content-Length response header before sending the headers (either explicitly via ngx.send_headers or implicitly via the first ngx.say or ngx.print call), then the HTTP 1.0 response buffering will be disabled even when this directive is turned on.
```

**Comment by agentzh** [comment]
@marcusramberg That is exactly what I was saying. Only `Content-Length` can affect the HTTP 1.0 buffering.

**Comment by marcusramberg** [comment]
@agentzh 'or implicitly via the first ngx.say og ngx.print call' Seems to indicate ngx.print should disable buffering. Am I not reading that right?

**Comment by agentzh** [comment]
@marcusramberg Implicit header sending triggered by ngx.say or ngx.print does not include a `Content-Length` response header at all. I agree that line can be ambiguous and needs tweaks.

**Comment by agentzh** [comment]
@marcusramberg After rereading that sentence, it is not ambiguous to me at all. You are required to set a `Content-Length` response header *explicitly* before header sending. And header sending is either an explicit `ngx.req.send_header` or an implicit `ngx.say`/`ngx.print` call. I don't see how you can get it wrong.

**Comment by agentzh** [comment]
@marcusramberg Neither `ngx.send_headers` nor `ngx.say`/ngx.print` can set a `Content-Length` header for you. You have to use `ngx.header["Content-Length"] = N` for it.
